src/ast/ast_validations.c

```diff
@@ -451,6 +478,90 @@ static AST_Validation _Validate_MATCH_Clause_Filters(const cypher_astnode_t *cla
 	return AST_VALID;
 }
 
+static AST_Validation _Validate_CALL_Clauses(const AST *ast, char **reason) {
+	/* Make sure procedure calls are valid:
+	 * 1. procedure exists
+	 * 2. number of arguments to procedure is as expected
+	 * 3. yield refers to procedure output */
+	AST_Validation res = AST_VALID;
+	ProcedureCtx *proc = NULL;
+	TrieMap *identifiers = NewTrieMap();
+
+	const cypher_astnode_t **call_clauses = AST_GetClauses(ast, CYPHER_AST_CALL);
+	if(call_clauses == NULL) return AST_VALID;
+
+	uint call_count = array_len(call_clauses);
+	for(uint i = 0; i < call_count; i ++) {
+		const cypher_astnode_t *call_clause = call_clauses[i];
+
+		// Make sure procedure exists.
+		const char *proc_name = cypher_ast_proc_name_get_value(cypher_ast_call_get_proc_name(call_clause));
+		proc = Proc_Get(proc_name);
+
+		if(proc == NULL) {
+			asprintf(reason, "Procedure `%s` is not registered", proc_name);
+			res = AST_INVALID;
+			goto cleanup;
+		}
+
+		// Validate num of arguments.
+		if(proc->argc != PROCEDURE_VARIABLE_ARG_COUNT) {
+			unsigned int given_arg_count = cypher_ast_call_narguments(call_clause);
+			if(Procedure_Argc(proc) != given_arg_count) {
+				asprintf(reason, "Procedure `%s` requires %d arguments, got %d", proc_name, proc->argc,
+						 given_arg_count);
+				res = AST_INVALID;
+				goto cleanup;
+			}
+		}
+
+		// Validate projections.
+		uint proj_count = cypher_ast_call_nprojections(call_clause);
+		if(proj_count > 0) {
+			// Collect call projections.
+			for(uint j = 0; j < proj_count; j++) {
+				const cypher_astnode_t *proj = cypher_ast_call_get_projection(call_clause, j);
+				const cypher_astnode_t *ast_exp = cypher_ast_projection_get_expression(proj);
+				assert(cypher_astnode_type(ast_exp) == CYPHER_AST_IDENTIFIER);
+				const char *identifier = cypher_ast_identifier_get_name(ast_exp);
+				TrieMap_Add(identifiers, (char *)identifier, strlen(identifier), NULL, TrieMap_DONT_CARE_REPLACE);
```

Comment:
We can check the return value of `TrieMap_Add` here to validate YIELD uniqueness, since constructions like `CALL db.labels() YIELD label, label` misbehave. In this case, Neo emits the error: "Variable `label` already declared".

Reply:
Brilliant